File: shell/src/lexer.c
#include "lexer.h"
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

static inline char at(Tokenizer *self) {
  if (self->pos >= self->code.len)
    return '\0';

  return self->code.data[self->pos];
}
static inline char eat(Tokenizer *self) {
  char prev = at(self);
  self->pos += 1;

  return prev;
}

static inline bool is_skippable(char x) {
  return x == ' ' || x == '\n' || x == '\t';
}
static inline bool is_eof(Tokenizer *self) {
  return self->pos >= self->code.len;
}

static inline Token make_token_with(Tokenizer *self, TokenType type,
                                    const char *lexeme) {
  size_t len = strlen(lexeme);
  uint8_t *start_ptr = (uint8_t *)lexeme;

  Str lexeme_slice = {.len = len, .data = start_ptr};
  Token token = {.type = type,
                 .lexeme = lexeme_slice,
                 .line = self->line,
                 .col = self->col};

  return token;
}
static inline Token make_token(Tokenizer *self, TokenType type, size_t start) {
  size_t len = self->pos - start - 1;
  uint8_t *start_ptr = self->code.data + start;

  Str lexeme = {.len = len, .data = start_ptr};
  Token token = {
      .type = type, .lexeme = lexeme, .line = self->line, .col = self->col};

  return token;
}

Token next(Tokenizer *self) {
  if (is_eof(self)) {
    return make_token_with(self, EOF, "<EOF>");
  }

  while (is_skippable(at(self))) {
    switch (at(self)) {
    case '0':
    case '\t':
      self->col += 1;
      self->pos += 1;
      break;
    case '\n':
      self->col = 0;
      self->line += 1;
      self->pos += 1;
      break;
    default:
      break;
    }
  }

  switch (at(self)) {
  default: {
    size_t start = self->pos;
    while (!is_skippable(at(self)) && !is_eof(self)) {
      eat(self);
    }
    return make_token(self, Normal, start);
  }
  }
}

File: Shell/src/Lexer.zig
const utils = @import("utils.zig");
const eql = utils.eql;
const Self = @This();

line: usize = 1,
column: usize = 0,
pos: usize = 0,
data: []const u8,

pub const Token = union(enum) {
    argument: []const u8,
    keyword: Keyword,
    eof,

    pub const Keyword = enum {
        @"const",
        @"var",
        @"fn",
        ret,
        @"if",
        @"else",
        @"while",
        @"for",
        @"break",
        @"continue",
        @"switch",
        @"or",
        @"and",
        not,
        end,
        @"dummy Keyword",

        const len = @intFromEnum(Keyword.@"dummy Keyword");
        /// comptime generates a keyword list, each keyword is an index into the returned value
        /// and the value is the keyword name as a string literal
        pub fn getKeywordlist() [len][]const u8 {
            var keywords: [len][]const u8 = .{""} ** len;

            inline for (keywords, 0..) |_, i| {
                const en: Keyword = @enumFromInt(i);
                keywords[i] = @tagName(en);
            }
            return keywords;
        }

        const keyword_list = getKeywordlist();
        pub fn fromString(str: []const u8) ?Keyword {
            for (keyword_list, 0..) |keyword, i| {
                if (eql(u8, keyword, str)) {
                    return @enumFromInt(i);
                }
            }
            return null;
        }
    };

    pub fn asString(self: *const Token) []const u8 {
        return switch (self.*) {
            .eof => "<EOF>",
            .keyword => @tagName(self.keyword),
            inline else => |token| token,
        };
    }

    pub fn debug(token: *const Token) !void {
        const print = @import("libc").stdio.zprintf;
        const string = token.asString();
        const tag = @tagName(token.*);
        try print("{%.*s: %.*s} ", .{ tag.len, tag.ptr, string.len, string.ptr });
    }
};

pub fn init(data: []const u8) Self {
    return .{ .data = data };
}

fn at(self: *const Self) u8 {
    return self.data[self.pos];
}

fn eat(self: *Self) u8 {
    defer self.pos += 1;
    defer self.column += 1;
    return self.at();
}

inline fn is_eof(self: *const Self) bool {
    return self.pos >= self.data.len;
}

inline fn is_skippable(self: *const Self) bool {
    const x = self.at();
    return x == '\n' or x == ' ' or x == '\t';
}

pub fn next(self: *Self) ?Token {
    while (!self.is_eof())
        switch (self.at()) {
            ' ', '\t' => _ = self.eat(),
            '\n' => {
                _ = self.eat();
                self.line += 1;
                self.column = 0;
            },
            else => break,
        };
    if (self.is_eof()) return null;
    const start = self.pos;

    while (!self.is_eof()) {
        if (!self.is_skippable()) _ = self.eat() else break;
    }

    const lexeme = self.data[start..self.pos];

    const keyword = Token.Keyword.fromString(lexeme) orelse return Token{ .argument = lexeme };
    return .{ .keyword = keyword };
}
